### ISSUE TITLE
Gap between block import and verification

### ISSUE THREAD

**Issue by phahulin** [paritytech/parity#7746]
> I'm running:
>
> - **Which Parity version?**: 1.8.4
> - **Which operating system?**: Linux
> - **How installed?**: binaries
> - **Are you fully synchronized?**: -
> - **Which network are you connected to?**: poa core
> - **Did you try to restart the node?**: yes

Consider the following scenario: there is a poa-type network with validator multi-set. A fork is planned at some future block (772000 in this case), so new element is added to the spec, like so:
```
"validators": {
  "multi": {
    "0": {
      "safeContract": "0x8bf38d4764929064f2d4d3a56520a76ab3df415b"
    },
    "772000": {
      "safeContract": "0x83451c8bc04d4ee9745ccc58edfab88037bc48cc"
    }
  }
}
```

fork goes smoothly at 772000 block
```
2018-01-29 17:27:10 UTC Initial contract validators: [cf260ea317555637c55f70e55dba8d5ad8414cb0, 3ab99ec45f706ee52748f24b1cea4676d91af64f, 8e6cdfacdae218ae312ad24cb1e8cf34bb9f6b61, 7a6a585db8cdfa88b9e8403c054ec2e912e9d32e]
2018-01-29 17:27:10 UTC Imported #772000 d907…5fd9 (0 txs, 0.00 Mgas, 1.53 ms, 0.57 KiB)
2018-01-29 17:27:15 UTC Imported #772001 1511…9caa (0 txs, 0.00 Mgas, 0.50 ms, 0.57 KiB)
2018-01-29 17:27:20 UTC Applying validator set change signalled at block 772000
```

On a node that did not update spec an error is produced:
```
2018-01-30 23:34:46  IO Worker #3 WARN client  Stage 5 block verification failed for #772003 (5fa3…bd0c)
Error: Block(InvalidStateRoot(Mismatch { expected: 1cf229025649b1a57a8ff1b783035200d8e8fc43136c71d0f0fc5d2c36542407, found: 5c0d62f8fe41ee96b5df6ea5f648b2ebd853d89742fe193e22c1c4fec67f1bee }))
```

Note, however that error happens at 77200**3** block. If spec is then replaced with correct one and parity restarted, it still can't sync, the error above reproduces. One has to delete chain files and resync from scratch. Could it be that some blocks are imported into local chain **before** they are verified?

**Comment by maxidev** [comment]
+1 same here.

**Comment by 5chdn** [comment]
>  Could it be that some blocks are imported into local chain before they are verified?

What most likely happened here is that the blocks #772000...002 are valid both for the old and new validator contract. Can you check that? If yes, this is actually desired behavior.

**Comment by phahulin** [comment]
How can I check for that?
I resynced with `--no-warp` using 1.10.6

1. removed 772000 hardfork from genesis file and started sync. Got the same error message:
```
2018-06-24 16:06:51  Syncing  #770264 8796…f068   839 blk/s    3 tx/s   0 Mgas/s      0+    0 Qed   #770264    3/25 peers   6 MiB chain 71 MiB db 0 bytes queue 5 MiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-06-24 16:06:53  Stage 5 block verification failed for #772003 (5fa3…bd0c)
Error: Block(InvalidStateRoot(Mismatch { expected: 1cf229025649b1a57a8ff1b783035200d8e8fc43136c71d0f0fc5d2c36542407, found: 5c0d62f8fe41ee96b5df6ea5f648b2ebd853d89742fe193e22c1c4fec67f1bee }))
2018-06-24 16:06:53  Imported #772002 0b64…51a4 (0 txs, 0.00 Mgas, 1.14 ms, 0.57 KiB) + another 5537 block(s) containing 33 tx(s)
2018-06-24 16:07:01  Syncing  #772002 0b64…51a4   174 blk/s    1 tx/s   0 Mgas/s      0+    0 Qed   #772001    2/25 peers   4 MiB chain 71 MiB db 0 bytes queue 17 KiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-06-24 16:07:11  Syncing  #772002 0b64…51a4     0 blk/s    0 tx/s   0 Mgas/s      0+    0 Qed   #772001    1/25 peers   4 MiB chain 71 MiB db 0 bytes queue 17 KiB sync  RPC:  0 conn,  0 req/s,   0 µs
```
this is expected

2. then I corrected genesis file and restarted parity - it still can't get past that block
```
2018-06-24 16:11:47  Syncing  #772002 0b64…51a4     0 blk/s    0 tx/s   0 Mgas/s      0+    0 Qed   #772002    1/25 peers   74 KiB chain 71 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-06-24 16:11:57  Syncing  #772002 0b64…51a4     0 blk/s    0 tx/s   0 Mgas/s      0+    0 Qed   #772002    2/25 peers   74 KiB chain 71 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-06-24 16:12:04  Stage 5 block verification failed for #772003 (5fa3…bd0c)
Error: Block(InvalidStateRoot(Mismatch { expected: 1cf229025649b1a57a8ff1b783035200d8e8fc43136c71d0f0fc5d2c36542407, found: 5c0d62f8fe41ee96b5df6ea5f648b2ebd853d89742fe193e22c1c4fec67f1bee }))
2018-06-24 16:12:12  Syncing  #772002 0b64…51a4     0 blk/s    0 tx/s   0 Mgas/s      0+    0 Qed   #771987    1/25 peers   74 KiB chain 71 MiB db 0 bytes queue 19 KiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-06-24 16:12:22  Syncing  #772002 0b64…51a4     0 blk/s    0 tx/s   0 Mgas/s      0+    0 Qed   #771747    1/25 peers   74 KiB chain 71 MiB db 0 bytes queue 73 KiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-06-24 16:12:32  Syncing  #772002 0b64…51a4     0 blk/s    0 tx/s   0 Mgas/s      0+    0 Qed   #771491    3/25 peers   74 KiB chain 71 MiB db 0 bytes queue 397 KiB sync  RPC:  0 conn,  0 req/s,   0 µs
```
Interestingly, block number in 11th column started to change

**Comment by 5chdn** [comment]
Let it run for a bit, it should reorg.

**Comment by phahulin** [comment]
It doesn't reorg. With time it looses all peers and disconnects from the network. I tried to restart, but the end-result is the same.

**Comment by 5chdn** [comment]
reset the DB and try again

**Comment by 5chdn** [comment]
can you check whether `#772002 0b64…51a4` is the same block as the `#772002` on the other chain?

**Comment by phahulin** [comment]
I'm not sure I understand what you want me to do. So here's what I did
1. synced from 0 to 772002 with incomplete genesis (missing fork at 772000). Checked hash of 772002 = 0x0b6439397610c4e9d2ed3abb6d1ef04bb470fe79be859c22ec372216dc3d51a4
2. replaced genesis file with correct one (including fork at 772000), restarted parity (not resynced!). Syncing doesn't continue past 772002 anyway, peers count eventually drops to 0. I checked hash of 772002 again - it's the same, but this result seems expected.
3. checked on the actual chain (height 3306274 now) - block 772002 have the same hash

So it looks like blocks 772000-772002 are correct in both cases (with and without fork), but it can't sync past 772002 even after genesis file is replaced with correct one
